pkg/components/prometheus-operator/component.go

```diff
@@ -120,10 +130,13 @@ func NewConfig() *component {
 			WatchLabeledServiceMonitors: true,
 			WatchLabeledPrometheusRules: true,
 		},
-		AlertManagerRetention:   "120h",
-		AlertManagerConfig:      defaultAlertManagerConfig,
-		AlertManagerStorageSize: "50Gi",
-		Namespace:               "monitoring",
+		AlertManager: &AlertManager{
+			Retention:   "120h",
+			Config:      defaultAlertManagerConfig,
+			StorageSize: "50Gi",
+		},
+		Namespace: "monitoring",
+		Operator:  &Operator{},
```

Comment:
We shouldn't need an empty object here.

Reply:
If not done it leads to invalid memory address or nil pointer dereference. It should not though, I don't know why this happens.

Reply:
I suggest the following patch:
```patch
diff --git a/pkg/components/prometheus-operator/component.go b/pkg/components/prometheus-operator/component.go
index 4aa47328..d648eb24 100644
--- a/pkg/components/prometheus-operator/component.go
+++ b/pkg/components/prometheus-operator/component.go
@@ -144,7 +144,6 @@ func NewConfig() *component {
                        StorageSize: "50Gi",
                },
                Namespace: "monitoring",
-               Operator:  &Operator{},
                Monitor: &Monitor{
                        Etcd:                  true,
                        KubeControllerManager: true,
@@ -226,14 +225,16 @@ func (c *component) RenderManifests() (map[string]string, error) {
                return nil, fmt.Errorf("rendering prometheus tolerations: %w", err)
        }

-       c.Operator.TolerationsRaw, err = util.RenderTolerations(c.Operator.Tolerations)
-       if err != nil {
-               return nil, fmt.Errorf("rendering operator tolerations: %w", err)
-       }
+       if c.Operator != nil {
+               c.Operator.TolerationsRaw, err = util.RenderTolerations(c.Operator.Tolerations)
+               if err != nil {
+                       return nil, fmt.Errorf("rendering operator tolerations: %w", err)
+               }

-       c.Operator.AdmissionWebhookTolerationsRaw, err = util.RenderTolerations(c.Operator.AdmissionWebhookTolerations) //nolint:lll
-       if err != nil {
-               return nil, fmt.Errorf("rendering operator admission webhook tolerations: %w", err)
+               c.Operator.AdmissionWebhookTolerationsRaw, err = util.RenderTolerations(c.Operator.AdmissionWebhookTolerations) //nolint:lll
+               if err != nil {
+                       return nil, fmt.Errorf("rendering operator admission webhook tolerations: %w", err)
+               }
        }

        c.AlertManager.TolerationsRaw, err = util.RenderTolerations(c.AlertManager.Tolerations)
diff --git a/pkg/components/prometheus-operator/template.go b/pkg/components/prometheus-operator/template.go
index 7b9d93f6..9d3d08d6 100644
--- a/pkg/components/prometheus-operator/template.go
+++ b/pkg/components/prometheus-operator/template.go
@@ -91,29 +91,24 @@ kubeEtcd:
   enabled: {{.Monitor.Etcd}}
 prometheus-node-exporter:
   service: {}
-{{ if (or .Operator.NodeSelector .DisableWebhooks .Operator.Tolerations ) }}
 prometheusOperator:
-  {{- if .DisableWebhooks }}
-  tlsProxy:
-    enabled: false
-  admissionWebhooks:
-    enabled: false
-  {{- end }}
-  {{- if .Operator.AdmissionWebhookTolerations }}
   admissionWebhooks:
+    enabled: {{ not .DisableWebhooks  }}
+  {{- with .Operator }}
+  {{- if .AdmissionWebhookTolerations }}
     patch:
-      tolerations: {{ .Operator.AdmissionWebhookTolerationsRaw }}
+      tolerations: {{ .AdmissionWebhookTolerationsRaw }}
   {{- end }}
-  {{- if .Operator.NodeSelector }}
+  {{- if .NodeSelector }}
   nodeSelector:
-    {{ range $key, $value := .Operator.NodeSelector }}
+    {{ range $key, $value := .NodeSelector }}
     {{ $key }}: {{ $value }}
     {{ end }}
   {{- end }}
-  {{- if .Operator.Tolerations }}
-  tolerations: {{ .Operator.TolerationsRaw }}
+  {{- if .Tolerations }}
+  tolerations: {{ .TolerationsRaw }}
+  {{- end }}
   {{- end }}
-{{ end }}
 prometheus:
   {{ if .Prometheus.Ingress }}
   ingress:
```